### PR TITLE
Implement Move Validation logic

### LINKED ISSUE #3: Move Validation Logic

### PR DESCRIPTION
**Pull request by RohrerF**
This PR resolves #3.

- [x] Implement Validation
  - [x] Calculate the board until that move
  - [x] Get the active boards
  - [x] Validate that the move is played on one of these active boards
  - [x] Validate that the move is played on a tile that is empty
- [x] Test validation
  - [x] CorrectPlayervalidation => start & after moves
  - [x] Board valid (cordinates valid / can play there)
  - [x] integrationtest => play a entire game

### CODE REVIEW

File: src/Domain/MoveValidator.cs [hunk 1]
```diff
@@ -8,36 +9,142 @@
...
+            // Winner ==> none
+        }
+
+        private bool IsBoardFull()
+        {
+            for (int x = 0; x < 3; x++)
+            {
+                for (int y = 0; y < 3; y++)
+                {
+                    if (_tiles[x, y].Value == TileValue.Empty)
```
Comment (panmona) [inline src/Domain/MoveValidator.cs]: Use list.Any instead?

Comment (RohrerF) [reply src/Domain/MoveValidator.cs]: The any extension is not available because _tiles is a multidimensional Array. In my opinion it's the only downside of using this type.

Comment (RohrerF) [reply src/Domain/MoveValidator.cs]: Switched Type. Any can be used now

File: src/Domain/TicTacToeGame.cs [hunk 2]
```diff
@@ -0,0 +1,202 @@
+using System;
```
Comment (RohrerF) [inline src/Domain/TicTacToeGame.cs:1]: @maracuja-juice Please review. The logic should be implemented. I will write tests soon to validate.

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (panmona) [inline .vscode/launch.json:1]: Not sure if we should check this folder in. These are user configurations.

File: src/Domain.Tests/BoardWinnerEvaluatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+{
+    public class BoardWinnerEvaluatorTests
+    {
+        [Theory]
+        [InlineData(Player.Circle)]
+        [InlineData(Player.Cross)]
+        private void GetWinner_EmptyBoard_ReturnsWinnerNone(Player lastPlayer)
+        {
+            // arrange
+            var emtyBoard =  new SmallTileInformation[3][];
```
Comment (panmona) [inline src/Domain.Tests/BoardWinnerEvaluatorTests.cs]: typo, should be empty

File: src/Domain/TicTacToeGame.cs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            {
+                if (lastMoveTilePosition.Equals(positionToValidate))
+                {
+                    return;
+                }
+
+                throw new IllegalPositionException();
+            }
+
+            if (_board[positionToValidate.X][positionToValidate.Y].Value == TileValue.Empty)
```
Comment (panmona) [inline src/Domain/TicTacToeGame.cs:122]: I think you should put this into a well named variable because I thought several minutes about why we need this, which is a sign that the code is not 100% clear.

File: src/Domain/TicTacToeGame.cs (hunk 2 above)
Comment (panmona) [reply src/Domain/TicTacToeGame.cs:1]: This looks great! I'm confident that it works correctly but yes, add some tests.

File: src/Domain.Tests/GameManagerTests.cs [hunk 6]
```diff
@@ -79,126 +95,50 @@ public async Task CreateGame_SavesGameInGameRepository()
         }
 
         [Fact]
-        public async Task Move_FirstMove_CorrectMoveNumber()
```
Comment (panmona) [inline src/Domain.Tests/GameManagerTests.cs:82]: Why was this removed?

**Review by panmona — DISMISSED** [review]
Looking great so far. ✨

File: src/Domain.Tests/GameManagerTests.cs (hunk 6 above)
Comment (RohrerF) [reply src/Domain.Tests/GameManagerTests.cs:82]: Figuring out the move number is no longer part of the GameManager

File: src/Domain/TicTacToeGame.cs (hunk 5 above)
Comment (RohrerF) [reply src/Domain/TicTacToeGame.cs:122]: any better name suggestions then nextBoardFromLastMove

Comment (panmona) [reply src/Domain/TicTacToeGame.cs:122]: Maybe we should use a different approach.
What was not clear for me is that we need this `if` for the case that the last board was full. (maybe it was also a bit too late when I looked at it)

File: src/Domain/TicTacToeGame.cs (hunk 2 above)
Comment (RohrerF) [reply src/Domain/TicTacToeGame.cs:1]: Tests show that it is somehow not working yet

Comment (RohrerF) [reply src/Domain/TicTacToeGame.cs:1]: The error seems to be fixed by implementing Equals on Position

Comment (panmona) [reply src/Domain/TicTacToeGame.cs:1]: :confetti_ball:

File: .vscode/launch.json (hunk 3 above)
Comment (RohrerF) [reply .vscode/launch.json:1]: It really helps beeing able to run the project in VSCode without having to write this stuff yourself

Comment (panmona) [reply .vscode/launch.json:1]: I don't have enough memory for VS Code so I didn't know that 😄

File: src/Abstractions.Tests/EnumExtensionsTests.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+namespace UltimateTicTacToe.Abstractions.Tests
+{
+    public class EnumExtensionsTests
+    {
+        [Theory]
+        [InlineData(Player.Circle, TileValue.Circle)]
+        [InlineData(Player.Cross, TileValue.Cross)]
+        public void ToTileValue_PlayerInput_ValidOutput(Player input, TileValue expectedOutput)
+        {
+            // arrange
```
Comment (panmona) [inline src/Abstractions.Tests/EnumExtensionsTests.cs:13]: Do we need these comments? They seem a bit unnecessary.
If yes: then at least make a space between `//` and `act`

File: src/Abstractions/Position.cs [hunk 8]
```diff
@@ -1,8 +1,30 @@
+using System;
+
 namespace UltimateTicTacToe.Abstractions
 {
-    public class Position
+    public sealed class Position : IEquatable<Position>
     {
+        public Position()
```
Comment (panmona) [inline src/Abstractions/Position.cs:7]: As far as I know, the C# compiler adds a parameterless constructor by default. So this isn't necessary.

File: src/Domain.Abstractions/Exceptions/ExceptionMessages.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    public static class ExceptionMessages
+    {
+        public const string InvalidInitialization =
+            "Not all moves used to initialize this game are valid. Move {0} is invalid";
+
+        public const string InvalidPlayer = "It's not the given Players turn";
+        public const string InvalidPosition = "The given position is not on the Board";
+        public const string IllegalPosition = "Not allowed to move here";
+        public const string TileNotEmpty = "Can't play here. Tile isn't empty";
+        public const string GameFinished = "Can't do any further moves. Game already finished";
```
Comment (panmona) [inline src/Domain.Abstractions/Exceptions/ExceptionMessages.cs]: Game is already finished 
(?)

File: src/Domain.Tests/TicTacToeGameTests.cs [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+
+            //act
+            Action a = () =>
+            {
+                new TicTacToeGame(new[]{new Move
+                {
+                    Player = Player.Cross,
+                    BoardPosition = new Position(-1,-1),
+                    TilePosition = new Position(-1,-1),
+                     MoveNumber = 1
```
Comment (panmona) [inline src/Domain.Tests/TicTacToeGameTests.cs]: formatting

File: src/Domain.Tests/TicTacToeGameTests.cs [hunk 11]
```diff
@@ -0,0 +1,508 @@
...
+                MoveNumber = 1
+            });
+
+            //assert
+            result.IsValid.Should().BeFalse();
+            result.InvalidReason.Should().Be(ExceptionMessages.InvalidPlayer);
+        }
+
+        [Fact]
+        public void Move_FistMoveByCross_ValidMove()
```
Comment (panmona) [inline src/Domain.Tests/TicTacToeGameTests.cs]: Fist? 🤛 👊 🤜

File: src/Domain.Tests/TicTacToeGameTests.cs [hunk 12]
```diff
@@ -0,0 +1,508 @@
...
+                    BoardPosition = new Position(1, 1),
+                    TilePosition = new Position(1, 0),
+                    MoveNumber = 16
+                },
+                new Move
+                {
+                    Player = Player.Cross,
+                    BoardPosition = new Position(1, 0),
+                    TilePosition = new Position(2, 1),
+                    MoveNumber = 16
```
Comment (panmona) [inline src/Domain.Tests/TicTacToeGameTests.cs]: 17

File: src/Domain.Tests/TicTacToeGameTests.cs [hunk 13]
```diff
@@ -0,0 +1,508 @@
...
+
+            var game = new TicTacToeGame(movesForFinishedGame);
+
+            //act
+            var result = game.Move(new Move
+            {
+                Player = Player.Circle,
+                BoardPosition = new Position(2, 1),
+                TilePosition = new Position(2, 1),
+                MoveNumber = 17
```
Comment (panmona) [inline src/Domain.Tests/TicTacToeGameTests.cs]: 18

File: src/Domain.Tests/TicTacToeGameTests.cs [hunk 14]
```diff
@@ -0,0 +1,508 @@
...
+                }
+            });
+
+            // act
+            var result = game.Move(new Move
+            {
+                Player = Player.Cross,
+                BoardPosition = new Position(2, 0),
+                TilePosition = new Position(2, 0),
+                MoveNumber = 2
```
Comment (panmona) [inline src/Domain.Tests/TicTacToeGameTests.cs]: 3

File: src/Domain.Tests/TicTacToeGameTests.cs [hunk 15]
```diff
@@ -0,0 +1,508 @@
...
+                    MoveNumber = 5
+                },
+            });
+
+            // act
+            var result = game.Move(new Move
+            {
+                Player = Player.Circle,
+                BoardPosition = new Position(0, 0),
+                TilePosition = new Position(0, 0),
```
Comment (panmona) [inline src/Domain.Tests/TicTacToeGameTests.cs]: I think this position should be on an empty tile for testing exactly what you want. 
Like: 1,1 or 2,2.
This tile was filled with Move 5.

**Review by panmona — APPROVED** [review]

**Review by panmona — CHANGES_REQUESTED** [review]
Some small changes needed. See comments.
Otherwise this is AmAzing! ✨😉😃

File: src/Abstractions.Tests/EnumExtensionsTests.cs (hunk 7 above)
Comment (RohrerF) [reply src/Abstractions.Tests/EnumExtensionsTests.cs:13]: I think we should have comments to make the Tripple A pattern visible. I've added the space after `//`

File: src/Abstractions/Position.cs (hunk 8 above)
Comment (RohrerF) [reply src/Abstractions/Position.cs:7]: Not if you have a constructor with value.

File: src/Domain.Abstractions/Exceptions/ExceptionMessages.cs (hunk 9 above)
Comment (RohrerF) [reply src/Domain.Abstractions/Exceptions/ExceptionMessages.cs]: fixed the typo

File: src/Domain.Tests/TicTacToeGameTests.cs (hunk 15 above)
Comment (RohrerF) [reply src/Domain.Tests/TicTacToeGameTests.cs]: good point. changed it

File: src/Abstractions/Position.cs (hunk 8 above)
Comment (panmona) [reply src/Abstractions/Position.cs:7]: Ah, you're right.

**Review by panmona — APPROVED** [review]
